### PR TITLE
feat: Add 'succeedModule' and 'stillValidModule' hooks

### PR DESCRIPTION
**Pull request by annanShao**
## Related issue (if exists)

[#3286 ](https://github.com/web-infra-dev/rspack/issues/3286)

### <samp>🤖 Generated by Copilot at c5f194f</samp>

This pull request adds two new hooks, `succeedModule` and `stillValidModule`, to the rspack web infrastructure, which allow plugins to handle module build and validity events. It implements these hooks in the Rust and JavaScript layers of the system, and adds test cases to verify their functionality. It also makes some minor changes to the types and traits related to the hooks to fix compiler errors and improve readability.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at c5f194f</samp>

*  Add two new hooks for plugins: `succeedModule` and `stillValidModule` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R29-R30), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R58-R59), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-f9c75b6b6bf1571ed8a920a1c40076219cb7006d1ade118a0f3af7a3dd7c84ceR28-R29), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1L19-R20), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R50-R51), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R482-R513), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R542-R543), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R621-R622), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-45ea00c1ab2a153d4a93ce2b5c307abd718b9613de4a66064e6379a82fdbf695R357-R360), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-3b15a372a6196f40438132383341257e7e52046aab993f8be03b053dc88e4c7eR538), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-3b15a372a6196f40438132383341257e7e52046aab993f8be03b053dc88e4c7eR546-R553), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1R91-R92), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1L124-R128), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL336-R338), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL595-R599), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdR725-R734))
  * The `succeedModule` hook is triggered when a module is successfully built ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R29-R30), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R58-R59), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-f9c75b6b6bf1571ed8a920a1c40076219cb7006d1ade118a0f3af7a3dd7c84ceR28-R29), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1L19-R20), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R50-R51), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R482-R513), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R542-R543), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R621-R622), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-45ea00c1ab2a153d4a93ce2b5c307abd718b9613de4a66064e6379a82fdbf695R357-R360), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-3b15a372a6196f40438132383341257e7e52046aab993f8be03b053dc88e4c7eR538), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1R91-R92), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1L124-R128), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL336-R338), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL595-R599), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdR725-R734))
    * Add a new variant to the `Hook` enum in `./crates/node_binding/src/hook.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R29-R30))
    * Add a new case to the `from_str` function in `./crates/node_binding/src/hook.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R58-R59))
    * Add a new field to the `JsHooks` struct in `./crates/node_binding/src/js_values/hooks.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-f9c75b6b6bf1571ed8a920a1c40076219cb7006d1ade118a0f3af7a3dd7c84ceR28-R29))
    * Add a new import, field, and method to the `PluginDriver` struct in `./crates/node_binding/src/plugins/mod.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1L19-R20), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R50-R51), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R482-R513))
    * Add a new argument and statement to the `new` function of the `PluginDriver` struct in `./crates/node_binding/src/plugins/mod.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R542-R543), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R591-R594))
    * Add a new argument and field to the `PluginDriver` struct initialization in `./crates/node_binding/src/plugins/mod.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R621-R622))
    * Add a new method to the `Plugin` trait in `./crates/rspack_core/src/plugin/api.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-45ea00c1ab2a153d4a93ce2b5c307abd718b9613de4a66064e6379a82fdbf695R357-R360))
    * Add a new attribute and method to the `PluginDriver` struct in `./crates/rspack_core/src/plugin/plugin_driver.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-3b15a372a6196f40438132383341257e7e52046aab993f8be03b053dc88e4c7eR538))
    * Add a new field and method to the `Compilation` class in `packages/rspack/src/compilation.ts` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1R91-R92), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1L124-R128))
    * Add a new field and method to the `Compiler` class in `packages/rspack/src/compiler.ts` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL336-R338), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdR725-R734))
    * Add a new field to the `hooks` object in the constructor of the `Compiler` class in `packages/rspack/src/compiler.ts` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL336-R338))
    * Add a new field to the `hookMap` object in the `#registerHooks` method of the `Compiler` class in `packages/rspack/src/compiler.ts` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL595-R599))
  * The `stillValidModule` hook is triggered when a module is valid and does not need to be rebuilt ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R29-R30), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R58-R59), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-f9c75b6b6bf1571ed8a920a1c40076219cb7006d1ade118a0f3af7a3dd7c84ceR28-R29), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-95c957e3bc52dfdce0db5276ec60213a6223a2dacce1659810127be853604b47L18-R18), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1L19-R20), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R50-R51), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R482-R513), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R542-R543), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R621-R622), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a98a14845029e2b2f9e0b9efc5c0fefe00d7d1c2e97ec1dbef9a4e2475745432L32-R32), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a98a14845029e2b2f9e0b9efc5c0fefe00d7d1c2e97ec1dbef9a4e2475745432L40-R40), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a98a14845029e2b2f9e0b9efc5c0fefe00d7d1c2e97ec1dbef9a4e2475745432L55-R55), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a98a14845029e2b2f9e0b9efc5c0fefe00d7d1c2e97ec1dbef9a4e2475745432L104-R104), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-5354812b8e30bff2ce2e8f7dfc292bdee3bad5022b1efc6fc1b85ab5a084eb14L232-R232), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-5354812b8e30bff2ce2e8f7dfc292bdee3bad5022b1efc6fc1b85ab5a084eb14L256-R269), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-45ea00c1ab2a153d4a93ce2b5c307abd718b9613de4a66064e6379a82fdbf695R357-R360), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-3b15a372a6196f40438132383341257e7e52046aab993f8be03b053dc88e4c7eR546-R553), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1R91-R92), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1L124-R128), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL336-R338), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL595-R599), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdR725-R734))
    * Add a new variant to the `Hook` enum in `./crates/node_binding/src/hook.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R29-R30))
    * Add a new case to the `from_str` function in `./crates/node_binding/src/hook.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-073c92c06772a11c0527a8a5c1e3780c11ed64ad8a8570e0d9a67bb13f856fc6R58-R59))
    * Add a new field to the `JsHooks` struct in `./crates/node_binding/src/js_values/hooks.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-f9c75b6b6bf1571ed8a920a1c40076219cb7006d1ade118a0f3af7a3dd7c84ceR28-R29))
    * Add a lifetime annotation to the `ToJsModule` trait in `./crates/node_binding/src/js_values/module.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-95c957e3bc52dfdce0db5276ec60213a6223a2dacce1659810127be853604b47L18-R18))
    * Add a new import, field, and method to the `PluginDriver` struct in `./crates/node_binding/src/plugins/mod.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1L19-R20), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R50-R51), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R482-R513))
    * Add a new argument and statement to the `new` function of the `PluginDriver` struct in `./crates/node_binding/src/plugins/mod.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R542-R543), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R591-R594))
    * Add a new argument and field to the `PluginDriver` struct initialization in `./crates/node_binding/src/plugins/mod.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-de65e35cd9f7b92416361921863c2eeb8b86694203ddd217395f16ce4e8164a1R621-R622))
    * Modify the return type and value of the `use_cache` method of the `BuildModuleOccasion` struct in `./crates/rspack_core/src/cache/occasion/build_module.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a98a14845029e2b2f9e0b9efc5c0fefe00d7d1c2e97ec1dbef9a4e2475745432L32-R32), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a98a14845029e2b2f9e0b9efc5c0fefe00d7d1c2e97ec1dbef9a4e2475745432L40-R40), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a98a14845029e2b2f9e0b9efc5c0fefe00d7d1c2e97ec1dbef9a4e2475745432L55-R55), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a98a14845029e2b2f9e0b9efc5c0fefe00d7d1c2e97ec1dbef9a4e2475745432L104-R104))
    * Modify the assignment of the `build_result` variable and add a new conditional statement to the `build_module` method of the `Queue` struct in `./crates/rspack_core/src/compiler/queue.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-5354812b8e30bff2ce2e8f7dfc292bdee3bad5022b1efc6fc1b85ab5a084eb14L232-R232), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-5354812b8e30bff2ce2e8f7dfc292bdee3bad5022b1efc6fc1b85ab5a084eb14L256-R269))
    * Add a new method to the `Plugin` trait in `./crates/rspack_core/src/plugin/api.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-45ea00c1ab2a153d4a93ce2b5c307abd718b9613de4a66064e6379a82fdbf695R357-R360))
    * Add a new method to the `PluginDriver` struct in `./crates/rspack_core/src/plugin/plugin_driver.rs` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-3b15a372a6196f40438132383341257e7e52046aab993f8be03b053dc88e4c7eR546-R553))
    * Add a new field and method to the `Compilation` class in `packages/rspack/src/compilation.ts` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1R91-R92), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-007fbe3e2cbba14328e3a87deaa9b6557a2d94e268cc07795c9a235a5b0575a1L124-R128))
    * Add a new field and method to the `Compiler` class in `packages/rspack/src/compiler.ts` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL336-R338), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdR725-R734))
    * Add a new field to the `hooks` object in the constructor of the `Compiler` class in `packages/rspack/src/compiler.ts` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL336-R338))
    * Add a new field to the `hookMap` object in the `#registerHooks` method of the `Compiler` class in `packages/rspack/src/compiler.ts` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-65c9f2211aa3417321f820555fd1ec4a9647fdc6cba3c80c2b96587a4d4ad4fdL595-R599))
* Add two test cases for the new hooks in `packages/rspack/tests/cases/hooks` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-fb03d98a375c8ff02cf5dc2509624eb8c19bd443daaf363a142e982ee863a372L1-R17), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-ca452c2ae7694ea3eb89586d225c44bcd6e6c9847eff13f9e9b6fa5b547950d2R1-R15), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-c0245cc5f0286007af50707d999a3af02be65e9e550b699e6f3ad9644cb1bf13R1-R13), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-2bd9114ff3583dda01275750ea11a1a61ce7fad00a5b7b9480c19f1e5ea0182fR1), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a7ec1f67988261ce36ee4dee801b4823090833e2385bbe1484168737de94edc7R1-R2), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-9f3f53b77f122a6f354c7157ee8e479e1161aafc8595fe7c20575096b9c49fd0L1-R17), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-6095ab2ec22054306a2e75dea527cecf961b80d9ecfba7b987472798445e2a5bR1-R12), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-78f83e80f848f0c3587d1a454ddf15a3fc98c84c4e658474c006f326b392a9a8R1-R14), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-e36dd5608511078d106c7d0052f0bdd2f8eb0740e2621a88defb54b301eec108R1), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-9ad9a538681b2d757c128b03b2a67c56d1a336447083f5d81c74be83e2e1921bR1-R2))
  * The `stillValidModule` test case tests the `stillValidModule` hook by logging a message and the module argument to the console when the hook is triggered ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-fb03d98a375c8ff02cf5dc2509624eb8c19bd443daaf363a142e982ee863a372L1-R17), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-ca452c2ae7694ea3eb89586d225c44bcd6e6c9847eff13f9e9b6fa5b547950d2R1-R15), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-c0245cc5f0286007af50707d999a3af02be65e9e550b699e6f3ad9644cb1bf13R1-R13), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-2bd9114ff3583dda01275750ea11a1a61ce7fad00a5b7b9480c19f1e5ea0182fR1), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a7ec1f67988261ce36ee4dee801b4823090833e2385bbe1484168737de94edc7R1-R2))
    * Add a `package.json` file for the test case package ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-fb03d98a375c8ff02cf5dc2509624eb8c19bd443daaf363a142e982ee863a372L1-R17))
    * Add a custom plugin that taps into the `stillValidModule` hook in `./plugins/MyStillValidModulePlugin.js` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-ca452c2ae7694ea3eb89586d225c44bcd6e6c9847eff13f9e9b6fa5b547950d2R1-R15))
    * Add a configuration file that sets the `context`, `mode`, `watch`, `entry`, and `plugins` fields for the compiler in `./rspack.config.js` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-c0245cc5f0286007af50707d999a3af02be65e9e550b699e6f3ad9644cb1bf13R1-R13))
    * Add a main module that imports a constant from another module and logs a message to the console in `./src/index.js` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-a7ec1f67988261ce36ee4dee801b4823090833e2385bbe1484168737de94edc7R1-R2))
    * Add a module that exports a constant in `./src/answer.js` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-2bd9114ff3583dda01275750ea11a1a61ce7fad00a5b7b9480c19f1e5ea0182fR1))
  * The `succeedModule` test case tests the `succeedModule` hook by logging a message and the module argument to the console when the hook is triggered ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-9f3f53b77f122a6f354c7157ee8e479e1161aafc8595fe7c20575096b9c49fd0L1-R17), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-6095ab2ec22054306a2e75dea527cecf961b80d9ecfba7b987472798445e2a5bR1-R12), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-78f83e80f848f0c3587d1a454ddf15a3fc98c84c4e658474c006f326b392a9a8R1-R14), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-e36dd5608511078d106c7d0052f0bdd2f8eb0740e2621a88defb54b301eec108R1), [link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-9ad9a538681b2d757c128b03b2a67c56d1a336447083f5d81c74be83e2e1921bR1-R2))
    * Add a `package.json` file for the test case package ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-9f3f53b77f122a6f354c7157ee8e479e1161aafc8595fe7c20575096b9c49fd0L1-R17))
    * Add a custom plugin that taps into the `succeedModule` hook in `./plugins/MySucceedModulePlugin.js` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-6095ab2ec22054306a2e75dea527cecf961b80d9ecfba7b987472798445e2a5bR1-R12))
    * Add a configuration file that sets the `context`, `mode`, `watch`, `entry`, and `plugins` fields for the compiler in `./rspack.config.js` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-78f83e80f848f0c3587d1a454ddf15a3fc98c84c4e658474c006f326b392a9a8R1-R14))
    * Add a main module that imports a constant from another module and logs a message to the console in `./src/index.js` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-9ad9a538681b2d757c128b03b2a67c56d1a336447083f5d81c74be83e2e1921bR1-R2))
    * Add a module that exports a constant in `./src/answer.js` ([link](https://github.com/web-infra-dev/rspack/pull/3318/files?diff=unified&w=0#diff-e36dd5608511078d106c7d0052f0bdd2f8eb0740e2621a88defb54b301eec108R1))

</details>

### CODE REVIEW

**Comment by hardfist** [comment]
!bench

**Comment by faga295** [comment]
https://github.com/web-infra-dev/rspack/blob/2526cd76f205075e3cdd358cc2386e0b45af20a4/packages/rspack/tests/case.template.ts#L53-L55
Maybe config file in test cases should named as `webpack.config.js`?

File: packages/rspack/tests/cases/hooks/stillValidModule/rspack.config.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const MyStillValidModulePlugin = require("./plugins/MyStillValidModulePlugin");
```
Comment (hardfist) [inline packages/rspack/tests/cases/hooks/stillValidModule/rspack.config.js:1]: better use webpack.config.js like other tests, I'm not sure whether out test template will load rspack.config.js

Comment (annanShao) [reply packages/rspack/tests/cases/hooks/stillValidModule/rspack.config.js:1]: got it

**Comment by IWANABETHATGUY** [comment]
Would you mind resolving the  conflicts?

**Comment by annanShao** [comment]
> Would you mind resolving the conflicts?

Yup I have resolved it right now.

**Comment by hardfist** [comment]
sorry can you resolve again, we will try merge ASAP

**Comment by annanShao** [comment]
> sorry can you resolve again, we will try merge ASAP

No problem, I have resolved it again, please check about it. Thanks.

**Review by hardfist — APPROVED** [review]

**Comment by hardfist** [comment]
can you help try whether eslint-webpack-plugin is valid now

**Comment by annanShao** [comment]
> can you help try whether eslint-webpack-plugin is valid now

Yup, I could test it later.

**Comment by annanShao** [comment]
It seems that eslint-webpack-plugin is working correctly now. I tested it using the eslint example (/examples/eslint).

**Comment by hardfist** [comment]
> It seems that eslint-webpack-plugin is working correctly now. I tested it using the eslint example (/examples/eslint).

good job

**Pull request merged**